File: R/regression.r
library(caTools)
library(ggplot2)

dataset = read.csv('../moar_data/Salary_Data.csv')

## Split the dataset into the training and test sets:
set.seed(123)
## sample.split(dependent_var (output), what_percentage_is_used_for_training)
split = sample.split(dataset$Salary, SplitRatio = 2/3) ## split the sample into the training and test, taking 80% of the cases as the training set.
## => Returns an array of true or false values.  True if that row was used for training, false if used for test.

training_set = subset(dataset, split == TRUE) ## Select all the data from the data set where it matches split at the same index as TRUE ?  I think?
test_set = subset(dataset, split == FALSE) ## Ditto, but where false.


## Simple linear regression

## Fit:
regressor = lm(
    formula = Salary ~ YearsExperience,
    data = training_set
)

## Predict:
y_pred = predict(regressor, new_data = test_set)

## Visualize the data:
ggplot() +
    geom_point(aes(x = training_set$YearsExperience, y = training_set$Salary),
               color = 'red') +
    geom_line(aes(x = training_set$YearsExperience, y = predict(regressor, new_data = training_set)),
              color = 'blue') +
    ggtitle('Salary vs XP (training)') +
    xlab('XP') +
    ylab('$$$')

File: python/regression.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import Imputer
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.cross_validation import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression


## Get the data:
dataset = pd.read_csv('../moar_data/Salary_Data.csv') ## Import the data.
X = dataset.iloc[:, :-1].values  ## Select All of the values (i.e. not numbers or headers) and then remove the last column (X is the independent var)
y = dataset.iloc[:, -1].values ## Get the dependent var (the right-most column)

## Split the dataset into the training set, and test set.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 1/3, random_state = 0)

## Simple linear regression:
regressor = LinearRegression()
regressor.fit(X_train, y_train)  ## This means:  "From the data I have provided calculate the line of best fit to all of the points on the scatter plot"

## Make the prediction:
y_pred = regressor.predict(X_test)  ## This means: "Given an x value, where does that x value intersect with the line I fitted to this data (on the y axis) i.e. what's the y value of the point on the line of best fit that matches this x value?"

## Visualize the data:
plt.scatter(X_train, y_train, color='red')
plt.scatter(X_test, y_pred, color='green')
plt.plot(X_train, regressor.predict(X_train), color='blue')
plt.title("Salary vs. XP (Training)")
plt.xlabel("XP")
plt.ylabel("$$$")
plt.show()
